keps/0017-pipe-tasks.md

```diff
@@ -0,0 +1,122 @@
+```yaml
+steps:
+    - name: bootstrap
+    tasks:
+        - gencert
+            - pipe:
+                - file: /usr/share/MyKey.key
+                  kind: Secret # ConfigMap
+                  key: {{.Pipes.Certificate}}
+``` 
+The key `{{.Pipes.Certificate}}` value will be generated by KUDO to avoid collisions and to make sure that it is not used before the file is generated. In the above example we would generate a secret name like `instance-myapp.deploy.bootstrap.gencert.pipes.certificate-#hash` to capture instance name along with plan/phase/step/task of the secret origin and the hash of its content. KUDO would use this hash to avoid recreating secrets for unchanged files on a plan rerun. We would also use labels ensuring that the secret is cleaned up when the corresponding Instance is removed.
+
+The corresponding `gencert` task can only have one resource definition:
+```yaml
+tasks:
+  gencert:
+    resources:
+      - gencert.yaml
+```
+
+For the pipe task resources, we only allow a [ContainerSpec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#container-v1-core) to be specified. Reasons for that are explained below in the implementation details. The ContainerSpec has to define a shared volume where the files are stored.
+
+```yaml
+spec:
+  volumes:
+  - name: shared-data
+    emptyDir: {}
+
+  containers:
+    - name: gencert
+      image: frapsoft/openssl
+      imagePullPolicy: Always
+      command: [ "sh", "-c" ]
+      args:
+        - openssl req -new -newkey rsa:4096 -x509 -sha256 -days 365 -nodes -out MyCertificate.crt -keyout /usr/share/MyKey.key
+      volumeMounts:
+        - name: shared-data
+          mountPath: /usr/share/
+```
+
+Any subsequent step resource (if the phase strategy is `serial`) might reference previously generated file by its key `{{.Pipes.Certificate}}` e.g.:
+```yaml
+spec:
+  containers:
+  - name: myapp
+    image: myapp:latest
+    volumeMounts:
+    - name: cert
+      mountPath: /еtc/certs
+  volumes:
+  - name: cert
+    secret:
+      secretName: {{.Pipes.Certificate}}
+```
+
+### Limitations
+- File generating container has to be side-effect free (meaning side-effects that are observable outside of the container like a 3rd party API call) as the container might be executed multiple times on failure. If that's not the case, `restartPolicy: Never` has to be set which would prevent the task from finishing successfully.
+- Only files <1Mb are applicable to be stored as ConfigMap or Secret.
+- Only one resource per pipe task is allowed. If needed multiple tasks must be used.
+
+### Implementation Details/Notes/Constraints
+
+There are several ways to implement pipe tasks, each one having its challenges and complexities. The approach below allows us not to worry about Pod container life-cycle as well as keep the storing logic in the KUDO controller:
```

Comment:
Is this for copying files in or copying files out? Init containers would work for copying in, but something else is needed for copying out. FWIW, Tekton takes an approach like you describe in alternatives to achieve this.

Reply:
Is it possible to kubectl copy from a completed pod?

Reply:
> Is it possible to kubectl copy from a completed pod?

No, it would not be completed. That's what the main `busybox` container with `sleep infinity` is for. Init container would generate files and complete but the main container would still run.  allowing us to copy out the files. I added a minimal working example demonstrating the proposed implementation.

Reply:
Thanks, the example helps. So the user's work will happen in the init container?

Reply:
Exactly. I know, it is somewhat counter-intuitive, but it is a normal `ContainerSpec` with slightly different life-cycle rules, which helps a lot in our case. Once the overall pod is `READY: 1/1, STATUS: Running` we know, that the init-container has run successfully and we can copy out files.

Reply:
Got it, that seems fine to me!